rust/examples-frontend/svc-simulcast/src/index.ts

```diff
@@ -0,0 +1,390 @@
+    id: ProducerId;
+}
+
+interface ServerConnectedConsumerTransport {
+    action: 'ConnectedConsumerTransport';
+}
+
+interface ServerConsumed {
+    action: 'Consumed';
+    id: ConsumerId;
+    kind: MediaKind;
+    rtpParameters: RtpParameters;
+}
+
+type ServerMessage =
+    ServerInit |
+    ServerConnectedProducerTransport |
+    ServerProduced |
+    ServerConnectedConsumerTransport |
+    ServerConsumed;
+
+interface ClientInit {
+    action: 'Init';
+    rtpCapabilities: RtpCapabilities;
+}
+
+interface ClientConnectProducerTransport {
+    action: 'ConnectProducerTransport';
+    dtlsParameters: DtlsParameters;
+}
+
+interface ClientConnectConsumerTransport {
+    action: 'ConnectConsumerTransport';
+    dtlsParameters: DtlsParameters;
+}
+
+interface ClientProduce {
+    action: 'Produce';
+    kind: MediaKind;
+    rtpParameters: RtpParameters;
+}
+
+interface ClientConsume {
+    action: 'Consume';
+    producerId: ProducerId;
+}
+
+interface ClientConsumerResume {
+    action: 'ConsumerResume';
+    id: ConsumerId;
+}
+
+interface ClientSetConsumerPreferredLayers {
+    action: 'SetConsumerPreferredLayers';
+    id: ConsumerId;
+    preferredLayers: {
+        spatialLayer: number
+        temporalLayer: number
+    }
+}
+
+type ClientMessage =
+    ClientInit |
+    ClientConnectProducerTransport |
+    ClientProduce |
+    ClientConnectConsumerTransport |
+    ClientConsume |
+    ClientConsumerResume |
+    ClientSetConsumerPreferredLayers;
+
+async function init()
+{
+    const sendPreview = document.querySelector('#preview-send') as HTMLVideoElement;
+    const receivePreview = document.querySelector('#preview-receive') as HTMLVideoElement;
+
+    sendPreview.onloadedmetadata = () => sendPreview.play();
+    receivePreview.onloadedmetadata = () => receivePreview.play();
+
+    const spatialLayerNode = document.querySelector('#spatial') as HTMLSpanElement
+    const temporalLayerNode = document.querySelector('#temporal') as HTMLSpanElement
+    const decreaseLayer = document.querySelector('#decreaseLayer') as HTMLButtonElement;
+    const increaseLayer = document.querySelector('#increaseLayer') as HTMLButtonElement;
+
+    let videoConsumer: Consumer | null = null;
+    let spatialLayers = 0;
+    let temporalLayers = 0;
+    let preferredSpatialLayer = 0;
+    let preferredTemporalLayer = 0;
+
+    decreaseLayer.addEventListener('click', () => {
+        let newPreferredSpatialLayer: number;
+        let newPreferredTemporalLayer: number;
+
+        if (preferredTemporalLayer > 0) {
+            newPreferredSpatialLayer = preferredSpatialLayer;
+            newPreferredTemporalLayer = preferredTemporalLayer - 1;
+        } else if (preferredSpatialLayer > 0) {
+            newPreferredSpatialLayer = preferredSpatialLayer - 1;
+            newPreferredTemporalLayer = temporalLayers - 1;
+        } else {
+            newPreferredSpatialLayer = spatialLayers - 1;
+            newPreferredTemporalLayer = temporalLayers - 1;
```

Comment:
Math on these lines seems to be incorrect

Reply:
why?
that's an implementation of mediasoup-demo as we agreed before

1) temporal can be decreased, we don't touch spatial
2) we try to decrease spatial because temporal is already 0.
3) we set max layers for both because they both were on 0.

Reply:
First branch here is correct, second branch is not because when it switches spatial layer down, it should switch temporal back to the highest value. Third branch simply allows layers to go into negative values instead of wrapping around.

Reply:
you misread the variable name... it doesn't go into a negative one...

Reply:
Ah, indeed. then `maxSpatialLayer` would be way less confusing than `spatialLayers - 1`. And similarly `increaseLayer` should use those variables instead of `2`.